### ISSUE TITLE
schedule automatic updates, restarts + announcements

### ISSUE THREAD

**Issue by edmundlaugasson** [linuxmint/mintupdate#651]
**Is your feature request related to a problem? Please describe.**
Problem is, that in case computers are working 24/7 and automatic updates are allowed, never don't know, when it really happens. Also if needed logout, restart, it is not announced to users. This all should be configurable. If computer restarted or shut down in the middle of updates, this can render Linux Mint into broken state.

**Describe the solution you'd like**
All the following options should be configurable and not mandatory.
Under Mintupdate there could be an option to allow automatic updates at fixed time. Usually in organizations there is certain time agreed for updates.
Also when update is running, don't allow do restart or shut down.
Allow to enable announcement (also custom message), when automatic updates are started and when finished.
Allow to enable showing also automatic updates progress in systray, visible as small icon, e.g. circle, that fulfills itself according automatic update progress.
When new kernel installed, announce about the need to restart computer and offer a button to restart right now.
Allow schedule automatic restart after automatic kernel or any other critical component installation. This should happen only, when new kernel or any other critical system component has been installed. People usually don't get any information and don't know to restart computer. Some people just wont restart computer regardless of appropriate messages. Again - it is agreed beforehand in organizations, that fixed time (w.g. after work, at night, etc) will be computers updated and if new kernel installed, also restarted. This will ensure, that automatically installed kernels and other critical components will be taken into use. Also automatic old kernel removal can work then.
There could be also automatic dynamical mode like in MS Windows, based on system usability activity. This applies, when user has very irregular times to work with computer, e.g. travelling often in different timezones, has just irregular working hours, etc.
Also it would be good to postpone automatic updates for certain time, e.g. offering previously configured time period, e.g. 5 minutes. Also user can enter the amount of time to postpone. Or offer also cancel for today or until certain date automatic updates. Although not postpone longer than certain number of days configured by administrator. This will ensure, that updates are still installed and Linux Mint will not get vulnerable.

**Describe alternatives you've considered**
Writing commands, script into CRON for updates, automatic restarts, announcements. Although determining new kernel installation after update and restart only then, has no good idea yet.
Announcements can be done using libnotify-bin package via CRON:
_/etc/update-messages.sh_
```
#!/bin/bash
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 20000 "Important message" "After 5 minutes, the software will update automatically and reboot the computer. Please save all your work and then close the programs you are using and leave your computer running."
/bin/sleep 1m
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send  -t 4000 "4 minutes" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 1m
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send  -t 4000 "3 minutes" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 1m
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send  -t 4000 "2 minutes" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 1m
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 4000 "1 minutes" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 5
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 4000 "55 seconds" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 5
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 4000 "50 seconds" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 5
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 4000 "45 seconds" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 5
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 4000 "40 seconds" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 5
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 4000 "35 seconds" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 5
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 4000 "30 seconds" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 5
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 4000 "25 seconds" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 5
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 4000 "20 seconds" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 5
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 4000 "15 seconds" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 5
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 4000 "10 seconds" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 5
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 4000 "5 seconds" "... then the software will be updated automatically and the computer will be rebooted"
/bin/sleep 5
export DISPLAY=:0; export XAUTHORITY=~/.Xauthority; /usr/bin/notify-send -t 10000000 "The software is being updated" "Please do nothing more with your computer. After that, the computer reboots automatically. Until the upgrade is complete - please do not use it. If there is a problem, please notify the administrator."
```

**Additional context**
Automatic reboots might sound not a good idea. Also allowing automatic updates with automatic old kernel removal in situation, where user will not do restarts due to lack of information about new kernels installed, is also not good idea. Allowing automatic updates with automatic new kernel installation inevitably requires automatic reboots in order to take newly installed kernels into effect. Certainly such activity requires prior agreement and this is not a problem. It is very common, that in organizations is fixed time for updates and if needed, also reboots.

**Comment by CTimmerman** [comment]
> Also if needed logout, restart, it is not announced to users.

I was about to file a new issue for this. The system still needing a reboot is pretty important to know, so use that red dot on the icon and add a "Reboot required." tooltip, just like you did for the nice welcome screen.

**Comment by clefebvre** [comment]
Reboots are inhibited during automated maintenance.

**Comment by breversa** [comment]
I came here from https://forums.linuxmint.com/viewtopic.php?p=2120554, because many times, my users tried to force-shutdown their computer while an automatic update was in progress, so all these would be great:

> - Allow to enable announcement (also custom message), when automatic updates are started and when finished. Announcing inside Mintupdate is not enough as in case of automatic updates, no Mintupdate window is open. Sounds like most efficient way is to show announcement balloon in system tray. Or consider other announcement ways.
> - Allow to enable showing also automatic updates progress in systray, visible as small icon, e.g. circle, that fulfills itself according automatic update progress.

**Comment by NostalgiaRealm** [comment]
I feel like this should get more attention. The update process should really be programmed to at the very least show us why a user cannot shut down the system. Or make it even more user friendly, so that when the user clicks the shut down button and finish updates either on next boot, or during shut down with an informative progress bar. Fedora has a progress bar like that, it informs the user. Unfortunately in the KDE spin of Fedora, automatic updates are not currently working for me, but that's something else.

Meanwhile, I have configured the unattendedupgrades with the MinimalSteps feature so apt updates are automated without any hassle. So far so good.

**Comment by svenportz77** [comment]
I strongly agree with NostalgiaRealm, breversa, CTimmerman and edmundlaugasson:
I love Linux Mint, I have used it since 2004 - and now I am installing and recommending it for my, mostly elderly, customers: theautomatic update function is of great help to those users.
Just recently I recognized with shock, what happens, if I force-shutdown (or enter my password to force-shotdown) during kernel-updates - I knew, what would happen, but I wanted to simulate an average/elderly customer: just shutdown the computer after using it, not knowing what runs in the background.... Of course, my laptop could not boot afterwards, saying "kernel-panic!".

So, such a notification, like all guys here mentioned, would be essential - and I hate to admit, but I have to honor Windows for showing a much clearer message.

As long as such a notification is missing, I will inform my customers. But because of their age they easily forget such things...
Else I have to stick with other distributions like MX Linux with KDE, which has such notifications built in, resp. applies downloaded updates after reboot/shutdown and switching on again.
Not my favorite choice, I do not like KDE too much, I like Cinnamon much more.

But better so then run after my customers, who broke their systems 😜
